compiler/rustc_query_system/src/query/mod.rs

```diff
@@ -25,31 +31,59 @@ use crate::dep_graph::{DepKind, DepNodeIndex, HasDepContext, SerializedDepNodeIn
 ///
 /// This is mostly used in case of cycles for error reporting.
 #[derive(Clone, Debug)]
-pub struct QueryStackFrame {
-    pub description: String,
-    span: Option<Span>,
-    pub def_id: Option<DefId>,
-    pub def_kind: Option<DefKind>,
-    /// A def-id that is extracted from a `Ty` in a query key
-    pub def_id_for_ty_in_cycle: Option<DefId>,
+pub struct QueryStackFrame<I> {
+    /// This field initially stores a `QueryStackDeferred` during collection,
+    /// but can later be changed to `QueryStackFrameExtra` containing concrete information
+    /// by calling `lift`. This is done so that collecting query does not need to invoke
+    /// queries, instead `lift` will call queries in a more appropriate location.
+    pub info: I,
```

Comment:
Does the distinction between the lazy and evaluated frame extra need to be encoded into type system?
It causes the new generic parameter to spread over all of the query system.

Can it be an enum (or a dyn trait like `LazyAttrTokenStream`) instead of a generic?
If this is only used in query cycles, then it probably won't affect the query system performance on a good path.

Reply:
To make it an enum, we'd need to add `'tcx` to the query system crate, so it could refer to `QueryStackDeferred`. That may spread further than the `I` parameter I used.